File: frontend/src/types/sequence.ts
export interface SequenceStep {
  day_offset: number
  subject_template: string
  body_template: string
  requires_approval: boolean
}

export interface Sequence {
  id: string
  name: string
  description: string | null
  steps: SequenceStep[]
  owner_user_id: string | null
  enabled: boolean
  created_at: number
  updated_at: number
}

export interface SequenceCreate {
  name: string
  description?: string | null
  steps: SequenceStep[]
  enabled?: boolean
}

export interface SequenceUpdate {
  name?: string
  description?: string | null
  steps?: SequenceStep[]
  enabled?: boolean
}

export interface Enrollment {
  id: string
  lead_id: string
  sequence_id: string
  enrolled_by_user_id: string
  step_idx: number
  status: string
  next_send_at: number | null
  started_at: number
  completed_at: number | null
}

export interface OutboundMessage {
  id: string
  enrollment_id: string
  lead_id: string
  user_id: string
  step_idx: number
  to_email: string
  subject: string
  body: string
  requires_approval: boolean
  status: string
  sent_at: number | null
  graph_message_id: string | null
  graph_conversation_id: string | null
  reply_at: number | null
  bounce_at: number | null
  error: string | null
  created_at: number
}

export interface PreviewRequest {
  sequence_id: string
  lead_id: string
  step_idx?: number
}

export interface PreviewResponse {
  subject: string
  body: string
  opener: string
  to_email: string | null
}

export interface BulkEnrollRequest {
  sequence_id: string
  lead_ids: string[]
}

export interface BulkEnrollResponse {
  enrolled: number
  skipped: number
}

File: backend/app/schemas/sequence.py
"""F3 — sequences, enrollments, outbound messages."""
from pydantic import BaseModel, Field


class SequenceStep(BaseModel):
    day_offset: int = 0
    subject_template: str
    body_template: str
    requires_approval: bool = False


class SequenceRead(BaseModel):
    id: str
    name: str
    description: str | None
    steps: list[dict]
    owner_user_id: str | None
    enabled: bool
    created_at: int
    updated_at: int

    model_config = {"from_attributes": True}


class SequenceCreate(BaseModel):
    name: str
    description: str | None = None
    steps: list[SequenceStep] = Field(default_factory=list)
    enabled: bool = True


class SequenceUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    steps: list[SequenceStep] | None = None
    enabled: bool | None = None


class EnrollmentRead(BaseModel):
    id: str
    lead_id: str
    sequence_id: str
    enrolled_by_user_id: str
    step_idx: int
    status: str
    next_send_at: int | None
    started_at: int
    completed_at: int | None

    model_config = {"from_attributes": True}


class BulkEnrollRequest(BaseModel):
    sequence_id: str
    lead_ids: list[str]


class BulkEnrollResponse(BaseModel):
    enrolled: int
    skipped: int


class OutboundMessageRead(BaseModel):
    id: str
    enrollment_id: str
    lead_id: str
    user_id: str
    step_idx: int
    to_email: str
    subject: str
    body: str
    requires_approval: bool
    status: str
    sent_at: int | None
    graph_message_id: str | None
    graph_conversation_id: str | None
    reply_at: int | None
    bounce_at: int | None
    error: str | None
    created_at: int

    model_config = {"from_attributes": True}


class PreviewRequest(BaseModel):
    sequence_id: str
    lead_id: str
    step_idx: int = 0


class PreviewResponse(BaseModel):
    subject: str
    body: str
    opener: str
    to_email: str | None
